package test_field;

enum MessageType {	
    NIL = 0;
    M2 = 2;
    M3 = 3;
    M4 = 4;
}

message TopMessage {
    required MessageType type = 1;
    optional SubMessage2 s2 = 2;
    optional SubMessage3 s3 = 3;
    repeated SubMessage4 s4 = 4;

	repeated int32 rf = 6;
}

message SubMessage2 {
    required int32 a2 = 1;
    optional int32 b2 = 2;
    optional int32 c2 = 3;
    optional string d2 = 4;
}


message SubMessage3 {
    required int32 a3 = 1;
    optional int32 b3 = 2;
    optional string d3 = 4;
}

message SubMessage4 {
    required int32 a4 = 1;
    optional int32 b4 = 2;
    optional int32 c4 = 3;
}
